File: backend/src/main/java/km/cd/backend/community/domain/Post.java
package km.cd.backend.community.domain;

import jakarta.persistence.*;
import km.cd.backend.challenge.domain.Challenge;
import km.cd.backend.common.domain.BaseTimeEntity;
import km.cd.backend.user.domain.User;
import lombok.*;

import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "posts")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Post extends BaseTimeEntity {

  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  private Long id;

  private String title;
  private String content;

  @ManyToOne(fetch = FetchType.LAZY)
  @JoinColumn(name = "challenge_id")
  private Challenge challenge;

  @ManyToOne(fetch = FetchType.LAZY)
  @JoinColumn(name = "author_id")
  private User author;

  @Setter
  private String image;

  private Boolean isRejected;

  @OneToMany(mappedBy = "post", orphanRemoval = true)
  private final List<Comment> comments = new ArrayList<>();

  @OneToMany(mappedBy = "post", orphanRemoval = true)
  private final List<Like> likes = new ArrayList<>();

  @Builder
  public Post(String title, String content, Challenge challenge, User author) {
    this.title = title;
    this.content = content;
    this.challenge = challenge;
    this.author = author;
    this.isRejected = false;
  }

  public boolean isAuthor(Long userId) {
    return this.author.getId().equals(userId);
  }

  public void rejectCertification() {
    this.isRejected = true;
  }

}

File: frontend/lib/model/data/post/post.dart
class Post {
  final int id;
  final String title;
  final String content;
  final String author;
  final int authorId;
  final String avatar;
  final String image;
  final String createdDate;
  final List<Comment> comments;
  final List<Like> likes;

  Post(
      {required this.id,
      required this.title,
      required this.content,
      required this.author,
      required this.authorId,
      required this.avatar,
      required this.image,
      required this.createdDate,
      required this.comments,
      required this.likes});

  factory Post.fromJson(Map<String, dynamic> json) {
    return Post(
      id: json['id'] as int,
      title: json['title'] as String,
      content: json['content'] as String,
      authorId: json['authorId'] as int,
      author: json['author'] as String,
      avatar: json['avatar'] as String,
      image: json['image'] as String,
      createdDate: json['createdDate'] as String,
      comments: (json['comments'] as List<dynamic>)
          .map((commentJson) =>
              Comment.fromJson(commentJson as Map<String, dynamic>))
          .toList(),
      likes: (json['likes'] as List<dynamic>)
          .map((likeJson) => Like.fromJson(likeJson as Map<String, dynamic>))
          .toList(),
    );
  }

  @override
  String toString() {
    return 'Post{id: $id, title: $title, content:  $content, author:  $authorId, $author, createdDate: $createdDate, image: $image, comments: $comments, likes: $likes}';
  }
}

class Comment {
  final int id;
  final int? parentId;
  final String author;
  final String avatar;
  final String content;
  final String createdDate;
  final List<Comment> children;

  Comment({
    required this.id,
    required this.author,
    required this.content,
    required this.avatar,
    required this.children,
    required this.createdDate,
    this.parentId,
  });

  factory Comment.fromJson(Map<String, dynamic> json) {
    return Comment(
        id: json['id'] as int,
        parentId: json['parentId'] as int?,
        author: json['author'] as String,
        content: json['content'] as String,
        avatar: json['avatar'] as String,
        children: json['children'] == null
            ? []
            : (json['children'] as List<dynamic>)
                .map((commentJson) =>
                    Comment.fromJson(commentJson as Map<String, dynamic>))
                .toList(),
        createdDate: json['createdDate'] as String);
  }
}

class Like {
  final int userId;

  Like({required this.userId});

  factory Like.fromJson(Map<String, dynamic> json) {
    return Like(
      userId: json['userId'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'userId': userId,
    };
  }
}
